core/src/main/clojure/xtdb/stagnant_log_flusher.clj

```diff
@@ -0,0 +1,111 @@
+            [xtdb.indexer]
+            [xtdb.tx-producer]
+            [juxt.clojars-mirrors.integrant.core :as ig])
+  (:import (java.nio.channels ClosedByInterruptException)
+           (java.time Duration)
+           (java.util.concurrent ExecutorService Executors ThreadFactory TimeUnit)
+           (xtdb.indexer IIndexer)
+           (xtdb.tx_producer ITxProducer)))
+
+;; see https://github.com/xtdb/xtdb/issues/2548
+
+(defonce thread-counter (atom -1))
+
+(defmethod ig/prep-key ::flusher
+  [_ opts]
+  (merge {:tx-producer (ig/ref :xtdb.tx-producer/tx-producer)
+          :indexer (ig/ref :xtdb/indexer)}
+         opts))
+
+(defmethod ig/init-key ::flusher
+  [_ {:keys [^ITxProducer tx-producer
+             ^IIndexer indexer
+             duration
+             ;; callback hook used to control timing in tests
+             ;; receives a map of the :last-flush, :last-seen tx-keys
+             on-heartbeat]
+      :or {duration #time/duration "PT4H"
+           on-heartbeat (constantly nil)}}]
+  (let [exr-tf (reify ThreadFactory
+                 (newThread [_ r]
+                   (doto (Thread. r)
+                     (.setName (str "xtdb.stagnant-log-flusher" (swap! thread-counter inc))))))
+        exr (Executors/newSingleThreadScheduledExecutor exr-tf)
+
+        ;; the tx-key of the last seen completed-tx
+        !last-seen-tx-id (atom nil)
+        ;; the tx-key of the last flush msg sent by me
+        !last-flush-tx-id (atom nil)
+
+        ^Runnable f
+        (bound-fn heartbeat []
+          (on-heartbeat {:last-flush @!last-flush-tx-id, :last-seen @!last-seen-tx-id})
+          (when-some [{latest-tx-id :tx-id} (.latestCompletedTx indexer)]
+            (try
+              (when (and (= @!last-seen-tx-id latest-tx-id)
+                         (or (nil? @!last-flush-tx-id)
+                             (< @!last-flush-tx-id latest-tx-id)))
+                (log/infof "log remains on tx-id %s, flushing any pending chunks" latest-tx-id)
+                (let [{flush-tx-id :tx-id} @(.submitTx tx-producer [[:flush-index-if-log-not-moving latest-tx-id]] {})]
+                  (log/debugf "flush tx = %s" flush-tx-id)
+                  (reset! !last-flush-tx-id flush-tx-id)))
+              (catch InterruptedException _)
+              (catch ClosedByInterruptException _)
+              (catch Throwable e
+                (log/error e "exception caught submitting :flush-index-if-log-not-moving transaction"))
+              (finally
+                (reset! !last-seen-tx-id latest-tx-id)))))]
+    {:executor exr
+     :task (.scheduleAtFixedRate exr f 0 (.toMillis ^Duration duration) TimeUnit/MILLISECONDS)}))
+
+(defmethod ig/halt-key! ::flusher [_ {:keys [^ExecutorService executor, task]}]
+  (future-cancel task)
+  (.shutdown executor)
+  (let [timeout-secs 10]
+    (when-not (.awaitTermination executor timeout-secs TimeUnit/SECONDS)
+      (log/warnf "flusher did not shutdown within %d seconds"))))
+
+(comment
```

Comment:
nit: worth shifting this to `dev`?

Reply:
Not exactly sure what you mean, could you clarify?

Reply:
we have a `dev` namespace with this kind of code - there are a few system definitions in there that folks can use to test certain things. would this system config be appropriate for standalone-config, for example? or maybe a separate config map?

Reply:
Oh yea, the sys config is difficult. The problem with sub-configurations is there is a large number of them I suppose. I suspect it's easier to start with a full node config and then partially configure it.

Reply:
Something like: 

```clojure
(def sys
    (-> (node/node-system {::flusher {:duration #time/duration "PT5S"}})
        (ig/prep)
        (ig/init [::flusher])))
```

Reply:
But I'd need to add the system constructor to do the defaulting and so on that node/start-node does